Improved coverage for unboxed->boxed kernel wrappers

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42054 add Dimname support to IValue
* **#38999 Improved coverage for unboxed->boxed kernel wrappers**

Adds boxing for inplace and outplace kernels, and traps
remaining unsupported cases. Fails compilation when
new unsupported types are introduced in op signatures.

See comments in [boxing.h](https://github.com/pytorch/pytorch/pull/38999/files#diff-63afcafe5c0813e43103b1a242566847) for details - TLDR, we use partial specialization to partition the space of kernel signatures (according to their arguments' boxing behavior and in/out/functional properties) and generate an appropriate unboxed-to-boxed wrapper for each kind.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

Commit:
Improved coverage for unboxed->boxed kernel wrappers

Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one...
3. the current rev doesn't identify unsupported types by name. May add that, but meanwhile the lvalue-reference logic should be ready for review.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. the current rev doesn't identify unsupported types by name in `TORCH_ASSERT`s thrown at runtime. May re-add those, but meanwhile the lvalue-reference logic should be ready for review.
4. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. the current rev doesn't identify unsupported types by name in `TORCH_ASSERT`s thrown at runtime. May re-add those, but meanwhile the lvalue-reference logic should be ready for review.
4. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. the current rev doesn't identify unsupported types by name in `TORCH_ASSERT`s thrown at runtime. May re-add those, but meanwhile the lvalue-reference logic should be ready for review.
4. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -7,73 +7,252 @@
+
+// is_tuple_of_lvalue_refs
+//
+template<class T, bool = is_specialization_of<std::tuple, T>::value>
```
not sure what the `bool = ...` is supposed to do. SFINAE? That only works if you have an `enable_if`.

bhosmer replied:
it has the same effect (diverting non-tuples to the other specialization) but I'll recast it in the more traditional form

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -7,73 +7,252 @@
+
+// tuple_elements
+//
+template <class Tuple, size_t... ns>
```
`tuple_elements` and `tuple_take` seem generic enough to warrant being somewhere more central, maybe `c10/util/Metaprogramming.h`, and get some unit tests in `Metaprogramming_test.cpp`.

bhosmer replied:
agree! done

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -7,73 +7,252 @@
+using can_unbox =
+  guts::conjunction<
+    guts::disjunction<
+      // using IValue(T) as a proxy for IValue.to<T>() here
```
you could also check for IValue.to<T> directly by creating a type trait with `void_t`.

```cpp
template<class T, class Enable = void> struct has_ivalue_to : std::false_type {};
template<class T> struct has_ivalue_to<T, guts::void_t<decltype(IValue::to<T>())>> : std::true_type {};
```

bhosmer replied:
ran into an unexpected wrinkle here, leaving it as is for now but will follow up

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -7,73 +7,252 @@
+
+// base definition should never be instantiated.
+// A "no call method defined on BoxedKernelWrapper" compile error means that
+// an op signature has failed to trigger any of the partial specialiations
```
nit: typo

bhosmer replied:
😬

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -7,73 +7,252 @@
+    Args... args
+  ) {
+    // TODO Reuse stack vector instead of allocating?
+    torch::jit::Stack stack;
```
as (very) recently discussed in the group chat, we might want to change this to

```cpp
torch::jit::Stack stack {std::forward<Args>(args)...};
```

to make sure `reserve()` is called correctly and the vector doesn't grow incrementally.

Same for the other implementations below.

bhosmer replied:
doing this provoked a use-before-inference type error in the `if_constexpr` following that had no obvious cause or solution... so a) I just reserved the capacity and b) it reinforced my feelings about if_constexpr being kind of fragile and difficult

bhosmer replied:
for posterity: turns out the problem is simply `IValue` not having a constructor for `TensorOptions`.

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -7,73 +7,252 @@
-  torch::jit::Stack stack;
-  torch::jit::push(stack, std::forward<Args>(args)...);
+struct BoxedKernelWrapper<Result(Args...), std::enable_if_t<has_quantizer_arg<Args...>::value, void>> {
+  static Result call(KernelFunction::InternalBoxedKernelFunction*, OperatorKernel*, const OperatorHandle&, Args... args) {
```
Have you tried a version where those unsupported cases also use `if_constexpr` instead of SFINAE? Is it more readable?

bhosmer replied:
tried it but, but it wasn't more readable. also, these should go away as support is added

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -7,73 +7,252 @@
+    OtherArgs... otherArgs
+  ) {
+    // TODO Reuse stack vector instead of allocating?
+    torch::jit::Stack stack;
```
hm this has some non-trivial changes to the above version, so folding it in with `if_constexpr` might not make sense. If we can figure out how to do it, however, then you could get rid of all your explaining comments about which version does what, they're only there because SFINAE is generally hard to understand IIUC and an if/else might not need them

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -7,73 +7,252 @@
+struct BoxedKernelWrapper<
+  Result(Result, OtherArgs...),
+  std::enable_if_t<
+    can_box_all<Result, OtherArgs...>::value && std::is_lvalue_reference<Result>::value,
```
are we sure we want to support the pattern that generically? I think we only need it for `Tensor&` right now, not for other lvalue references, and by keeping only `Tensor&` supported, we prevent people from starting to use it for other types.

bhosmer replied:
agree, restricted it to Tensor&

smessmer commented on aten/src/ATen/templates/TensorMethods.cpp:
```diff
@@ -25,7 +25,7 @@ namespace at {
 // we'll remove them when we are actually exposing Quantizer class
 // to frontend
-using ConstQuantizerPtr = const c10::intrusive_ptr<Quantizer>&;
+// using ConstQuantizerPtr = const c10::intrusive_ptr<Quantizer>&;
```
should we delete this instead of commenting it out?

bhosmer replied:
😬

bhosmer commented on aten/src/ATen/core/boxing/KernelFunction_test.cpp:
```diff
@@ -43,6 +44,45 @@ void boxed_func_without_return(const OperatorHandle& /*opHandle*/, Stack* stack)
+  // (Tensor(a!), Scalar) -> Tensor(a!)
+  EXPECT_EQ(2, stack->size());
+
+  EXPECT_TRUE(stack->at(0).isTensor());
```
ah good catch, thanks

bhosmer commented on aten/src/ATen/core/boxing/KernelFunction_test.cpp:
```diff
@@ -147,6 +234,46 @@ void expectUnboxedCallingWithoutReturnWorks(const KernelFunction& func) {
+  at::Tensor& t = func.call<at::Tensor&, at::Tensor&, at::Scalar>(dummy, a, 1.0f);
+
+  EXPECT_EQ(a.item().toFloat(), 1.0f);
+  EXPECT_EQ(t.item().toFloat(), 1.0f);
```
here yes, but the tuple case raises a question that actually affects both... will follow up offline

bhosmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -7,73 +7,252 @@
+// is_specialization_of
+//
+template <template <typename...> class T, typename U>
+struct is_specialization_of: std::false_type {};
```
oh good, using that

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

ezyang commented:
At a high level, I still would love a more in depth PR description here. One metric by which I judge PR descriptions is that I would like to be able, to some degree, "predict" what code changes would be in the PR from the description.  "Adds boxing for inplace and outplace kernels" says "what" but it doesn't really say how we went about doing this, and as a reviewer the how is helpful because it will help me understand what the code is doing as I review it.

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -5,75 +5,247 @@
+
+// is_tensor_ref
+template<class T> struct is_tensor_ref : std::false_type {};
+template <> struct is_tensor_ref<at::Tensor&> : std::true_type {};
```
nit: this would be clearer if you say it's a mutable ref. In the C++ standard library, "reference" canonically refers to both const and non-const references, e.g., `static_assert(std::is_reference<const int&>());`

bhosmer replied:
definitely

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -5,75 +5,247 @@
+// that follow.
+//
+template<class FuncType, class Enable = void>
+struct BoxedKernelWrapper {};
```
pro-tip: You could also put a static assert like `static_assert(sizeof(FuncType) == -1, "msg")` to customize the error message when partial specialization fails

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -5,75 +5,247 @@
+//
+// The partial specializations below handle various cases: in
+// particular, not all types appearing in op signatures are supported,
+// and ops returning references have nonstandard wrapper implementations.
```
This comment goes a long way to explaining what I was missing from the PR description

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -5,75 +5,247 @@
+// These specializations capture the remaining gaps in boxing support.
+// Rather than triggering compile errors, we generate boxed kernels that
+// raise runtime errors. As support for these types is added, the
+// specializations can be removed.
```
Nice!

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -5,75 +5,247 @@
+  ) {
+    // TODO Reuse stack vector instead of allocating?
+    torch::jit::Stack stack;
+    stack.reserve(sizeof...(Args));
```
In some cases certain argument types convert into multiple stack entries (`TensorOptions` is the most notable example):

```
static inline void push_one(Stack& stack, c10::TensorOptions options) {
  stack.emplace_back(c10::typeMetaToScalarType(options.dtype()));
  stack.emplace_back(options.layout());
  stack.emplace_back(options.device());
  stack.emplace_back(options.pinned_memory());
}
```

In those cases, the reservation computed using `sizeof...` would be too small. Beware!

bhosmer replied:
😮 

Excellent point. We could notice TensorOptions in the args and over-reserve in that case (absent an exact member count), but I think we want to try to go to zero allocations rather than keeping it to just 1, so I'm less inclined to deal with the TO case until we try to move the whole thing to the machine stack, or TLS, or whatever. 

But definitely worth noting that stack size >= argc. cc @smessmer @ljk53 re IValue stack tactics

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -5,75 +5,247 @@
+
+    return guts::if_constexpr<!std::is_same<void, Result>::value>(
+      [&] (auto delay_check) {
+        TORCH_INTERNAL_ASSERT(
```
Probably should be debug only asserts

bhosmer replied:
ah yeah totally.

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -5,75 +5,247 @@
+          stack.size() == 1,
+          "Boxed kernel was expected to push exactly one return value to the stack."
+        );
+        return delay_check(std::move(stack[0]).to<Result>());
```
TIL. (I didn't know how to use the delay check idiom for return types. Now I know.....)

bhosmer replied:
Ahahaha yeah well basically YIL... but calling it `delay_check` [instead of `_`](https://github.com/pytorch/pytorch/blob/master/c10/util/C++17.h#L242-L251) definitely helps me read the idiom 😬

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -5,75 +5,247 @@
+//
+
+// A boxable arg type is one that IValue has a constructor for.
 // Assume T is decayed
```
There is a nasty little asymmetry here, which is that `can_box` assumes the inputs are decayed, but `can_unbox` arguments MUST NOT be decayed (because if they are, then you can't identify `Tensor&` as a return type that cannot be unboxed.)

bhosmer replied:
Excellent point! The "assume decayed" always felt fragile to me, but I didn't feel like I had a good enough reason to move it into the body. UNTIL NOW

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -5,75 +5,247 @@
+//
+template<class Result, class... OtherArgs>
+struct BoxedKernelWrapper<
+  Result(Result, OtherArgs...),
```
Suggestion for more clarity: I think you can pattern match directly on `Tensor&` here.

So instead of testing `is_tensor_ref<Result>`, you just have `Tensor&(Tensor&, OtherArgs...)` here. I think the net result should be the same.

bhosmer replied:
😍

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -5,75 +5,247 @@
+    torch::jit::push_one(stack, outArg);
+    torch::jit::push(stack, std::forward<OtherArgs>(otherArgs)...);
+
+    (*boxed_kernel_func)(functor, opHandle, &stack);
```
Since we didn't change the internal calling convention, if I understand correctly, usually there will be some goop left on the stack (specifically, an `outArg` entry) once we're done calling the boxed kernel function, is that right? Seems like a decent oppo for a debug only assert.

bhosmer replied:
yeah, good idea, also for the tuple-returning case

ezyang reviewed (approved):
The partial specializations were very readable and I was able to understand the code on the first read through, thank you!

ezyang commented:
I heard that malfet may have removed the naughty windows CI config. However, if you find you still need to work around:
 
```
..\aten\src\ATen/core/boxing/impl/boxing.h(116): error C2210: '_Ty': pack expansions cannot be used as arguments to non-packed parameters in alias templates
..\aten\src\ATen/core/boxing/impl/boxing.h(131): error C2210: '_Ty': pack expansions cannot be used as arguments to non-packed parameters in alias templates
..\aten\src\ATen/core/boxing/impl/boxing.h(141): error C2210: '_Ty': pack expansions cannot be used as arguments to non-packed parameters in alias templates
..\aten\src\ATen/core/boxing/impl/boxing.h(187): error C2210: '_Ty': pack expansions cannot be used as arguments to non-packed parameters in alias templates
```

My guess is that converting our alias templates (the templated `using` declarations) into honest to goodness structs would also be sufficient to work around the MSVC bug.

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

Commit:
Update on "Improved coverage for unboxed->boxed kernel wrappers"


Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Reviewer notes:
1. `QuantizerBase.h` is just the classdef for `Quantizer` snipped out of `Quantizer.h` and placed in its own file, so `Quantizer` can be fully defined in `TensorMethods.h` without introducing a circularity. (The definition was not previously needed bc the relevant ops didn't get autoboxing support.)
2. there are some general utilities at the top of `boxing.h` which don't comfortably fit in any of the existing headers, but I didn't want to make a new one... is there a better home for them?
3. Best to leave custom class boxing tests out of this PR, I think.

Differential Revision: [D21718547](https://our.internmc.facebook.com/intern/diff/D21718547)

[ghstack-poisoned]

bhosmer commented:
> I heard that malfet may have removed the naughty windows CI config. However, if you find you still need to work around:

@ezyang looks like the new windows config fixes it, but thank you for the suggestions!

ilia-cher commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -353,14 +353,9 @@ inline Return Dispatcher::callWithDispatchKey(const TypedOperatorHandle<Return(A
         guard.before(op.schema().name(), stack, at::sequence_number::peek());
-
-        // if we could convert all the arguments, also pass the stack into the kernel call
-        if (boxed_all_args) {
-          return kernel.template callBoxedOrUnboxed<Return, Args...>(op, stack, std::forward<Args>(args)...);
```
just curious, why did we decide to remove this branch?

bhosmer replied:
@ilia-cher sorry, meant to comment on this here myself, thanks for flagging! Eliminating `callBoxedOrUnboxed` saves a nontrivial amount of codesize - ~160k in libtorch.so release build. Separately, letting the control flow fall through to the unboxed `call` - rather than using `callBoxed` here and early returning - was more of a judgement call*, and one we can revisit after [complete boxing support](https://github.com/pytorch/pytorch/pull/42438/files#diff-355e2c36e6a2cc684d9341da5eed62f8R354) lands.

*about what percentage of the kernels we'll call from here will be unboxed

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -68,61 +103,195 @@ inline bool boxArgumentsOrCannotBoxIntoStack(std::vector<c10::IValue>& stack, co
-  // case, just call them anyway without a stack.  These special cases can be
-  // removed once we support boxing everything.
-  // See Note [named_not_supported_kernel]
-  if (boxed_kernel_func == &named_not_supported_kernel) {
```
@bhosmer I just realized that we lost this logic entirely. It is not a problem to lose this logic when we 100% support boxing, since it is obsolete at that point, but prior to that, missing this logic means that you'll get a much worse error message when you hit one of the special `named_not_supported_kernel` on a codepath that doesn't support boxing it. I don't recall how close to 100% boxing we are now so I don't know the severity of this.

bhosmer replied:
We're at 100% boxing now (mod reordering issue, but all types are supported) 😁  but thanks for the callout! I thought about this earlier today as well, can't remember what triggered it - something in your presentation maybe?
